src/fsharp/vs/ServiceFormatting/CodePrinter.fs

```diff
@@ -0,0 +1,1144 @@
+﻿// Copyright (c) Microsoft Corporation.  All Rights Reserved.  See License.txt in the project root for license information.
+
+module internal Microsoft.FSharp.Compiler.SourceCodeServices.ServiceFormatting.CodePrinter
+
+open System
+open Microsoft.FSharp.Compiler.Ast
+open FormatConfig
+open SourceParser
+open SourceTransformer
+
+/// This type consists of contextual information which is important for formatting
+type ASTContext =
+    {
+      /// Original file name without extension of the parsed AST 
+      TopLevelModuleName: string 
+      /// Current node is the first child of its parent
+      IsFirstChild: bool
+      /// Current node is a subnode deep down in an interface
+      IsInterface: bool 
+      /// This pattern matters for formatting extern declarations
+      IsCStylePattern: bool
+      /// Range operators are naked in 'for..in..do' constructs
+      IsNakedRange: bool
+      /// The optional `|` in pattern matching and union type definitions
+      HasVerticalBar: bool
+      /// A field is rendered as union field or not
+      IsUnionField: bool
+      /// First type param might need extra spaces to avoid parsing errors on `<^`, `<'`, etc.
+      IsFirstTypeParam: bool
+      /// Check whether the context is inside DotGet to suppress whitespaces
+      IsInsideDotGet: bool
+    }
+    static member Default =
+        { TopLevelModuleName = "" 
+          IsFirstChild = false; IsInterface = false 
+          IsCStylePattern = false; IsNakedRange = false
+          HasVerticalBar = false; IsUnionField = false
+          IsFirstTypeParam = false; IsInsideDotGet = false }
+
+let rec addSpaceBeforeParensInFunCall functionOrMethod arg = 
+    match functionOrMethod, arg with
+    | _, ConstExpr(Const "()", _) -> false
+    | SynExpr.LongIdent(_, LongIdentWithDots s, _, _), _ ->
+        let parts = s.Split '.'
+        not <| Char.IsUpper parts.[parts.Length - 1].[0]
+    | SynExpr.Ident(Ident s), _ -> not <| Char.IsUpper s.[0]
+    | SynExpr.TypeApp(e, _, _, _, _, _, _), _ -> addSpaceBeforeParensInFunCall e arg
+    | _ -> true
+
+let addSpaceBeforeParensInFunDef functionOrMethod args =
+    match functionOrMethod, args with
+    | _, PatParen (PatConst(Const "()", _)) -> false
+    | "new", _ -> false
+    | (s:string), _ -> 
+        let parts = s.Split '.'
+        not <| Char.IsUpper parts.[parts.Length - 1].[0]
+    | _ -> true
+
+let rec genParsedInput astContext = function
+    | ImplFile im -> genImpFile astContext im
+    | SigFile si -> genSigFile astContext si
+
+and genImpFile astContext (ParsedImplFileInput(hs, mns)) = 
+    col sepNone hs genParsedHashDirective +> (if hs.IsEmpty then sepNone else sepNln)
+    +> col sepNln mns (genModuleOrNamespace astContext)
```

Comment:
This is not a common operator right? We should add a note to get rid of it eventually

Reply:
Yeah, it's defined here 

https://github.com/Microsoft/visualfsharp/pull/3542/files/e2b56b340d5f5e70d55e9b777a596fa88439f79b#diff-f003beb5354b3fc47c79aaaf3e23a7c7R199

as "function composition".